### ISSUE TITLE
GvrAudio linker problems with Xcode 8.3.2

### ISSUE THREAD

**Issue by oa2013** [googlevr/gvr-unity-sdk#559]
Hello! I've upgraded to Xcode 8.3.2 recently and now my Unity project build fails with the following errors (it looks like a bunch of GvrAudio functions are lacking the _GvrAudio prefix):

`Undefined symbols for architecture arm64:
  "_SetSourceGain", referenced from:
      _GvrAudio_UpdateAudioSource_m910492288 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSourceGain_m2700692092 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetSourceGain_m2700692092)
  "_SetSourceDistanceAttenuationMethod", referenced from:
      _GvrAudio_UpdateAudioSource_m910492288 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSourceDistanceAttenuationMethod_m2494613792 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetSourceDistanceAttenuationMethod_m2494613792)
  "_CreateSource", referenced from:
      _GvrAudio_CreateAudioSource_m1984649488 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_CreateSource_m3333004860 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_CreateSource_m3333004860)
  "_DestroySoundfield", referenced from:
      _GvrAudio_DestroyAudioSoundfield_m3923888282 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_DestroySoundfield_m1843937534 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_DestroySoundfield_m1843937534)
  "_SetSoundfieldGain", referenced from:
      _GvrAudio_UpdateAudioSoundfield_m1370440907 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSoundfieldGain_m2065600452 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetSoundfieldGain_m2065600452)
  "_SetSoundfieldRotation", referenced from:
      _GvrAudio_UpdateAudioSoundfield_m1370440907 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSoundfieldRotation_m1542399254 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetSoundfieldRotation_m1542399254)
  "_SetSourceSpread", referenced from:
      _GvrAudio_UpdateAudioSource_m910492288 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSourceSpread_m1252472104 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetSourceSpread_m1252472104)
  "_SetSourceTransform", referenced from:
      _GvrAudio_UpdateAudioSource_m910492288 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSourceTransform_m2386695377 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetSourceTransform_m2386695377)
  "_SetListenerTransform", referenced from:
      _GvrAudio_UpdateAudioListener_m1154610187 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetListenerTransform_m592380371 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetListenerTransform_m592380371)
ld: symbol(s) not found for architecture arm64
clang: error: linker command failed with exit code 1 (use -v to see invocation)
`
I've tried to use both GVR SDK **1.3.0** and **1.0.3** and the result is the same. All the library references to GVR SDK in Xcode look correct.  

Is there a workaround that would allow me to build my project?

**Comment by oa2013** [comment]
Update:  I was able to build my project by leaving the following setting under Unity's build options **blank**:

<img width="316" alt="screen shot 2017-05-08 at 8 55 40 pm" src="https://cloud.githubusercontent.com/assets/5304739/25831434/102c3f6c-3432-11e7-9e51-54583eb09593.png">

I'm not sure what point Unity started auto-populating this field. Running Unity 5.5.3f1 here.

**Comment by guneyozsan** [comment]
I also get similar errors with both Unity 5.6.0f3 and Unity 2017.1.0b4 with GVR 1.40. However I can build with Unity 5.6.0f3 and GVR 1.30 with some custom modifications.

Unfortunately @oa2013's solution to set Scripting Define Symbols doesn't work for me. The error I get is:
`
Undefined symbols for architecture armv7:
  "_Initialize", referenced from:
      _GvrAudio_Initialize_m2739768702 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_Initialize_m2441973946 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _CameraFader_Initialize_m1628979316, _VRUIManager_InitializeFlow_m3589339368 , _VRHud_InitializeHUD_m2274019903_MetadataUsageId , _MACAlgorithm_Initialize_m133885420 , _VRSimulationManager_InitializeSimulation_m2451371400_MetadataUsageId , __Z26CMVideoSampling_InitializeP15CMVideoSampling , _GvrAudioSoundfield_InitializeSoundfield_m2000343270_MetadataUsageId , _Reporter_Initialize_m1735167020 , _SHA384Managed_Initialize_m1719088674 , _ObjectPool_1_Initialize_m4283417319_gshared , _GameObjectPoolController_Initialize_m1707273073 , _VRSelectPlayMenu_InitializePlaysList_m498272043 , _GameObjectPoolController_Initialize_m1707273073_MetadataUsageId , _GvrAudio_Initialize_m2739768702 , _GvrAudioSoundfield_InitializeChannelSet_m2342420071 , _ObjectPool_1_Initialize_m756943073_RuntimeMethod_var , _Reporter_Initialize_m1735167020_MetadataUsageId , _VRUIManager_InitializeFlow_m3589339368_MetadataUsageId , _XmlTextWriter_Initialize_m467029607 , _GvrAudio_Initialize_m2441973946 , _GvrAudioSoundfield_InitializeSoundfield_m2000343270 , _GvrAudioSource_InitializeSource_m594433190 , _SHA256Managed_Initialize_m2091245045 , _SHA1Managed_Initialize_m3538700609 , _SHA1Internal_Initialize_m2776411059 , _PersistentCallGroup_Initialize_m3908067069_MetadataUsageId , _XmlTextWriter_Initialize_m467029607_MetadataUsageId , _GvrAudioSoundfield_InitializeChannelSet_m2342420071_MetadataUsageId , _GvrAudioSource_InitializeSource_m594433190_MetadataUsageId , _GvrAudio_Initialize_m2739768702_MetadataUsageId , _SortSequenceContext_2_Initialize_m3487377048_gshared , _XmlInputStream_Initialize_m630070605_MetadataUsageId , _RuntimeHelpers_InitializeArray_m3920580167 , _VRSimulationManager_InitializeSimulation_m2451371400 , _VRHud_InitializeHUD_m2274019903 , _XmlTextReader_InitializeContext_m3588031662_MetadataUsageId , _UriParser_InitializeAndValidate_m251943319 , _VRUIManager_InitializePreSnapIndicatorStatus_m1666174004 , _VRSelectPlayMenu_InitializePlaysList_m498272043_MetadataUsageId , _VRUIManager_InitializePreSnapIndicatorStatus_m1666174004_MetadataUsageId , _MD2Managed_Initialize_m1330748427 , _ObjectPool_1_Initialize_m4283417319_MetadataUsageId , _XmlSerializationWriter_Initialize_m2848121882_MetadataUsageId , _SortSequenceContext_2_Initialize_m2437603572_gshared , _StreamReader_Initialize_m1533993200 , _MACTripleDES_Initialize_m1318099802_MetadataUsageId , _SHA384Managed_Initialize_m712692291 , _SHA1CryptoServiceProvider_Initialize_m3685014747 , _UnitySynchronizationContext_InitializeSynchronizationContext_m2728554850_MetadataUsageId , _DirectoryInfo_Initialize_m2216118894_MetadataUsageId , _RIPEMD160Managed_Initialize_m3365407578 , _RuntimeHelpers_InitializeArray_m2578109054 , _CryptoConfig_Initialize_m1553095052_MetadataUsageId , _SHA512Managed_Initialize_m891287980 , _HMAC_Initialize_m3052649104 , _MD4Managed_Initialize_m2316730601 , _MD5SHA1_Initialize_m3440022970 , _CipherSuite_InitializeCipher_m229247572 , _XmlSerializationReader_Initialize_m2808791690_MetadataUsageId , _SslHandshakeHash_Initialize_m2786849785 , _BlockProcessor_Initialize_m3476609078 , _SortKeyBuffer_Initialize_m588272701 , _SortKeyBuffer_Initialize_m588272701_MetadataUsageId , _HMAC_Initialize_m2760853000_MetadataUsageId , _Array_Initialize_m3086099726 , _XmlSerializationReaderInterpreter_InitializeList_m1067879528 , _LayoutRebuilder_Initialize_m2671585490 , _DirectoryInfo_Initialize_m2216118894 , _StreamWriter_Initialize_m1707678624 , _StreamReader_Initialize_m1533993200_MetadataUsageId , _StreamWriter_Initialize_m1707678624_MetadataUsageId , _RuntimeHelpers_InitializeArray_m3920580167_MetadataUsageId , _CryptoConfig_Initialize_m1553095052 , _HMAC_Initialize_m2760853000 , _MACTripleDES_Initialize_m1318099802 , _MD5CryptoServiceProvider_Initialize_m486876542 , _SHA512Managed_Initialize_m2899358361 , _XmlTextReader_InitializeContext_m3588031662 , _XmlSerializationReader_Initialize_m2808791690 , _XmlSerializationWriter_Initialize_m2848121882 , _XmlInputStream_Initialize_m630070605 , _FMOD_Memory_Initialize , _PersistentCallGroup_Initialize_m3908067069 , _UriParser_InitializeAndValidate_m251943319_MetadataUsageId , _UnitySynchronizationContext_InitializeSynchronizationContext_m2728554850 , __Z45Register_UnityEngine_Network_InitializeServerv , __Z87Register_UnityEngine_Experimental_Rendering_DrawRendererSettings_InitializeSortSettingsv , __Z37CustomRenderTexture_CUSTOM_InitializeP12Il2CppObject , __Z42PlayerConnectionInternal_CUSTOM_Initializev , __Z33Network_CUSTOM_InitializeSecurityv , _OBJC_IVAR_$_AVFoundationMediaLoader.m_Initialized , _UnityDisplayManager_Initialize , __Z35GUI_CUSTOM_InitializeGUIClipTexturev , __Z50Network_CUSTOM_Internal_InitializeServerDeprecatedii , __Z47Register_UnityEngine_Network_InitializeSecurityv , __Z49Register_UnityEngine_GUI_InitializeGUIClipTexturev , __Z31Network_CUSTOM_InitializeServeriih , __Z64Register_UnityEngine_Network_Internal_InitializeServerDeprecatedv , __Z56Register_UnityEngine_PlayerConnectionInternal_Initializev , __Z51Register_UnityEngine_CustomRenderTexture_Initializev , __Z50DrawRendererSettings_CUSTOM_InitializeSortSettingsP12Il2CppObjectP24DrawRendererSortSettings )
  "_SetSourceBypassRoomEffects", referenced from:
      _GvrAudio_UpdateAudioSoundfield_m2589129101 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_SetSourceBypassRoomEffects_m1691460578 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_UpdateAudioSource_m1447747246 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_SetSourceBypassRoomEffects_m1691460578)
  "_Shutdown", referenced from:
      _GvrAudio_Shutdown_m2838331605 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_Shutdown_m3814448932 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_Shutdown_m2838331605_MetadataUsageId, _GvrAudioSource_ShutdownSource_m801851364_MetadataUsageId , _GvrAudioSource_ShutdownSource_m801851364 , _GvrAudioSoundfield_ShutdownChannelSet_m1300044223 , _GvrAudio_Shutdown_m2838331605 , _GvrAudio_Shutdown_m3814448932 , _GvrAudioSoundfield_ShutdownSoundfield_m2554412452_MetadataUsageId , _GvrAudioSoundfield_ShutdownSoundfield_m2554412452 , _Socket_Shutdown_internal_m3459802987 , __Z20FMOD_OS_Net_Shutdownv , __ZN3Geo16GeoMemoryDefault11ms_ShutdownE )
  "_CreateSoundfield", referenced from:
      _GvrAudio_CreateAudioSoundfield_m1783995047 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_CreateSoundfield_m278242392 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_CreateSoundfield_m278242392)
  "_SetListenerGain", referenced from:
      _GvrAudio_UpdateAudioListener_m3370947442 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_SetListenerGain_m1924947456 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_SetListenerGain_m1924947456)
  "_CreateSoundObject", referenced from:
      _GvrAudio_CreateAudioSource_m1984649488 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_CreateSoundObject_m1329629085 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_CreateSoundObject_m1329629085)
  "_SetSourceListenerDirectivity", referenced from:
      _GvrAudio_UpdateAudioSource_m1447747246 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_SetSourceListenerDirectivity_m1770884030 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_SetSourceListenerDirectivity_m1770884030)
  "_DestroySource", referenced from:
      _GvrAudio_DestroyAudioSource_m801182782 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_DestroySource_m2875741446 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_DestroySource_m2875741446)
  "_SetSourceOcclusionIntensity", referenced from:
      _GvrAudio_UpdateAudioSource_m1447747246 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_SetSourceOcclusionIntensity_m2889416367 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_SetSourceOcclusionIntensity_m2889416367)
  "_SetSourceDirectivity", referenced from:
      _GvrAudio_UpdateAudioSource_m1447747246 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_SetSourceDirectivity_m2025518334 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_SetSourceDirectivity_m2025518334)
  "_SetRoomProperties", referenced from:
      _GvrAudio_UpdateAudioRoom_m3179955199 in Bulk_Assembly-CSharp_1.cpp.obj
      _GvrAudio_SetRoomProperties_m2331769136 in Bulk_Assembly-CSharp_1.cpp.obj
     (maybe you meant: _GvrAudio_SetRoomProperties_m2331769136)
ld: symbol(s) not found for architecture armv7
`

**Comment by miraleung** [comment]
Could you please try again with 1.40 of the GVR Unity SDK, and ensure that Assets/Plugins/iOS/libaudioplugingvrunity.a exists? If not, it needs to be imported from Assets/GoogleVR/GVRiOSNativeCompatibility.unitypackage.

**Comment by oa2013** [comment]
Hi Mira!

The plot thickens. I took out GVR 1.3 out of my project and imported 1.4. Unity now repeatedly crashes with the same error originating somewhere in libsystem_kernel.dylib (see the screenshot).

Other projects with GVR 1.3 and below do not exhibit the same behavior.  I'll install the Unity 5.6 version alongside 5.5 and will let you know if I get the same error message. 

![screen shot 2017-05-10 at 9 22 58 pm](https://cloud.githubusercontent.com/assets/5304739/25927958/7fae4b6e-35c7-11e7-8e8c-b7f4968c46f6.png)

**Comment by miraleung** [comment]
Deleting Library/ and Temp/ may fix certain types of unexplained crashes, especially those resulting from Unity version changes.

As for the undefined symbols, it looks as though a .a file is missing from Assets/Plugins/iOS.

**Comment by veerb4h** [comment]
Hi oa2013 
I am also facing same issue as you were facing but not able to fix it yet , can you suggested me better way to fix this issue

[Unity_Build_Error.txt](https://github.com/googlevr/gvr-unity-sdk/files/1007773/Unity_Build_Error.txt)

**Comment by oa2013** [comment]
Hey veerb4h,

What version of the GVR SDK and Unity and Xcode are you using? Here are my Unity settings for GVR 1.3 for Unity 5.5.3 for Xcode 8.3 (see the three attached screenshots):

<img width="458" alt="screen shot 2017-05-18 at 8 35 11 am" src="https://cloud.githubusercontent.com/assets/5304739/26210613/084a4c6a-3ba5-11e7-85fe-9b8d4e9b07fd.png">
<img width="454" alt="screen shot 2017-05-18 at 8 35 22 am" src="https://cloud.githubusercontent.com/assets/5304739/26210615/085729f8-3ba5-11e7-91f1-88dc38d24d11.png">
<img width="475" alt="screen shot 2017-05-18 at 8 35 28 am" src="https://cloud.githubusercontent.com/assets/5304739/26210614/0854afde-3ba5-11e7-965b-32580ad003c2.png">

The scripting symbols doesn't seem to make any difference. But this does produce a xcodeproject that Xcode doesn't complain about -- no GvrAudio errors or any other errors. 

I've not had a chance to install Unity 5.6 in parallel with 5.5.

**Comment by veerb4h** [comment]
Hi @oa2013 
I am using Unity 5.5.3 with GVR SDK v 1.3, but not able build my project for iOSApp still getting anothe  18 error
[Unity_iOS_Xcode_Build_Error.txt](https://github.com/googlevr/gvr-unity-sdk/files/1028486/Unity_iOS_Xcode_Build_Error.txt)

**Comment by oa2013** [comment]
Hi  @veerb4h ,

Can you please double-check that the Gvr Audio Spatializer is selected under Project Settings > Audio in Unity?

<img width="355" alt="screen shot 2017-05-26 at 6 39 32 pm" src="https://cloud.githubusercontent.com/assets/5304739/26515339/e9e8cc20-4247-11e7-8c17-eb9549cd2927.png">

**Comment by miraleung** [comment]
@oa2013 For the crash, could you please try deleting Temp/ and Library/, then reimporting the project?

As for the missing symbols error, v1.40 of the GVR Unity SDK has a separate iOS audio library package, which contains those missing symbols, that should be used with Unity 5.6.1p1. For v1.30, could you please see if there's a libgvrunity.a file in Assets/Plugins/iOS?

**Comment by oa2013** [comment]
Hi Mira @miraleung!

My stuff's compiling now (on Unity 5.5 with SDK 1.3). 

Deleting the folders you mentioned and reimporting fixed the missing symbols issue for me but it sounds like @veerb4h is having the same problem based on his errors and he followed the same steps.  

I'm not sure what else could be the issue except for not including the Gvr Audio Spatializer plugin. Ideas?

**Comment by miraleung** [comment]
@veerb4h Could you please try deleting Assets/GoogleVR, and the libgvrunity.a in Assets/Plugins/iOS, then replacing with a fresh copy of v1.30? It looks like some code has changed in there.

**Comment by oa2013** [comment]
Hi again @miraleung!

I've upgraded to Unity 5.6.1f1 and I'm getting the libPods-Unity-iPhone.a linker error with GVR v. 1.3.0.  Deleting Assets/GoogleVR and Assets/Plugins/iOS isn't making a difference -- I made sure I deletes and re-imported a few times for consistency.

Here's what I see in Xcode when Unity finishes building the project:

**1)** 
<img width="250" alt="screen shot 2017-06-06 at 8 13 38 pm" src="https://user-images.githubusercontent.com/5304739/26856947-0cdc32de-4af5-11e7-82e0-b1785722d066.png">

Since all these are missing, I delete them, try to build in Xcode and then get an error that says the **GVROverlayView.h** file is missing (it was in one of the subfolders under that Pods folder).

I put the file here and build again...

**2)** 
<img width="410" alt="screen shot 2017-06-06 at 8 19 03 pm" src="https://user-images.githubusercontent.com/5304739/26856984-66310044-4af5-11e7-9bcd-ad2a55a0a411.png">

**3)** Now, I get this GvrAudio error we've seen before: 
```
Undefined symbols for architecture arm64:
  "_SetSourceGain", referenced from:
      _GvrAudio_UpdateAudioSource_m910492288 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSourceGain_m2700692092 in Bulk_Assembly-CSharp_0.o
     (maybe you meant: _GvrAudio_SetSourceGain_m2700692092)
  "_SetSourceDistanceAttenuationMethod", referenced from:
      _GvrAudio_UpdateAudioSource_m910492288 in Bulk_Assembly-CSharp_0.o
      _GvrAudio_SetSourceDistanceAttenuationMethod_m2494613792 in Bulk_Assembly-CSharp_0.o

```
**4)** Mucking around with the Xcode linker flags hasn't helped so far, so I'll go back to Unity 5.5.3. 

Thanks for helping work out the kinks in the Unity/Xcode pipeline!

**Comment by miraleung** [comment]
Thanks @oa2013, we're investigating internally. In the meantime, have you seen the workaround in #504?

**Comment by oa2013** [comment]
Hey @miraleung, yes, I've tried the workaround mentioned in #504, deleting the $(inherited) flag from Xcode but no luck this time. It has worked before. 

I am back to Unity 5.5 for the moment but of course eventually I'll need to use GVR 1.5 and hence upgrade to Unity 5.6+. 

In a related note, I tried using Unity 2017.1 beta and it's not recognizing the Cardboard SDK at all but that's a Unity issue.

**Comment by miraleung** [comment]
Another tip that just came in: Please open the workspace, and not the project. We're looking into updating the Unity documentation.

**Comment by fredsa** [comment]
We've updated the [Get started with Google VR in Unity on iOS](https://developers.google.com/vr/unity/get-started-ios) guide. The new guide includes a note about the importance of opening the generated `*.xcworkspace` file instead of the `*.xcodeproj` file.

Closing this issue as I believe the above problems are fixed by a combination of:
* Updating to the latest Unity 5.6 or Unity 2017 release
* Updating to the latest Google VR SDK for Unity
* Opening the the generated `*.xcworkspace` Xcode workspace
* Following the steps in the above mentioned getting started guide

@oa2013  Please let us know (by filing a new issue) if you encounter any issues creating a working Xcode project using the latest Unity 5.6 or Unity 2017 and the latest Google VR SDK for Unity.
